### ISSUE TITLE
Linux Support

### ISSUE THREAD

**Issue by karasuhebi** [JamesWilko/Payday-2-BLT#56]
Is this planned? :D

**Comment by BangL** [comment]
as far as i know it wouldnt be that easy..
for now we are "hacking" the "IP Helper API", which seems to be windows-only.
so on linux, we would need a completely different concept. a simple "port" isnt possible.
so don't put your hopes too high. ;)

**Comment by karasuhebi** [comment]
Oh sorry I didn't mean to imply it'd be easy, I was pretty certain it would not be something trivial and would definitely require some work. Thanks for confirming my suspicions though. Here's to hoping it happens though! :)

**Comment by BangL** [comment]
well.. don't get me wrong. i am totally hoping the same!
too bad i have no idea about C/C++, otherwise i would try to help

**Comment by polarathene** [comment]
So something like [this](https://rafalcieslak.wordpress.com/2013/04/02/dynamic-linker-tricks-using-ld_preload-to-cheat-inject-features-and-investigate-programs/)? 

Could you list what would be involved/required for someone to tackle a linux hook. It looks like you are mapping the lua methods like briefly covered [here](https://sites.google.com/site/sbobovyc/writing/reverse-engineering/hooking-lua-part-1) for making a hook, what else would the linux version need? 

Besides the file stuff, how much of BLT's codebase would you say is shareable(platform independent)?

**Comment by minermoder27** [comment]
Wow. That actually looks fairly simple - just 'override' the calls used to the the Lua VM, and inject something to load out of user-defined files.

**Comment by SirWaddles** [comment]
I haven't looked at Payday's linux version just yet, but if it's anything like the Windows one, I don't think that the LD_PRELOAD environment variable will be adequate.

Given that Payday has its own version of the lua lib, I doubt it would be being used a shared library. So in the final compiled version, calls aren't being made via the function symbol, they're just going straight to the function address in the binary, so compiling a shared library with matching symbols I don't think will be able to intercept anything.

In all the examples I've found regarding the LD_PRELOAD trick, including the one that polarathene mentioned, only system calls are intercepted, things that are strictly shared libraries.

I hope I'm wrong, I've not looked into it extensively. I'd like to support linux, but I'm not really sure how.

**Comment by polarathene** [comment]
I have rather limited experience/knowledge in this area. I've had luck manipulating games with CheatEngine on Windows in the past, a common approach is to locate a static pointer address in memory to override a value or function(alternatively hook it).

If unable to do that because the target data is dynamic due to being created on the fly from something like Lua, you could try your luck with Array of Byte(AOB) scans. The latter is a method I used for Payday 2 before being aware of Lua hooks.

I imagine it's much harder to locate the actual Lua functions addresses in memory, and that they're likely to change with updates(AOB scans might help here). If the addresses in memory were found, would the trick I linked be viable, or the addresses useful for mapping/hooking in some other way? 

If this approach could work, how does compiling lua into an application affect it when loaded into memory, is there a chance of identifying/matching similarities of the version of lua used to another program that embeds lua into it's compiled application? Or would the binary data related to lua be completely different between the two compiled applications? If the former we might have a better chance of isolating function calls to match with PD2's lua in memory.

**Comment by minermoder27** [comment]
ldd on payday2_release (the linux executable)
no liblua.so.1 in sight...

linux-vdso.so.1
libGL.so.1
libopenal.so.1
libsteam_api.so
libm.so.6
libdl.so.2
libgcc_s.so.1
libpthread.so.0
libc.so.6
libnvidia-tls.so.352.63
libnvidia-glcore.so.352.63
libX11.so.6
libXext.so.6
librt.so.1
libstdc++.so.6
/lib64/ld-linux-x86-64.so.2
libxcb.so.1
libXau.so.6
libXdmcp.so.6

**Comment by SirWaddles** [comment]
Hey, while you're here, can you run the 'nm' utility on the binary? It should show what symbols have been exported. Probably do a pastebin dump or something, because there's likely to be a few.

**Comment by SirWaddles** [comment]
@polarathene We're already using signature scanning on the windows one. I highly doubt that the signatures will exactly match between platforms, so we'd have to find new sigs. As well as we'd need ways to find the process handle (using a windows.h function at the moment) and to replace some existing functions (using Microsoft Detours at the moment)

**Comment by minermoder27** [comment]
Ok, done the nm dump.
If you want to lag your browser, go to http://sprunge.us/JSiI to view the 4MB of result.

**Comment by SirWaddles** [comment]
@minermoder27 Hey, it's not a stripped binary. That solves a few problems.

**Comment by minermoder27** [comment]
Sorry, I didn't see your link. Run nm -an | c++filt

Done with those flags at http://sprunge.us/TQSU

**Comment by SirWaddles** [comment]
That just unmangles the c++ stuff, it's basically the same either way.

**Comment by minermoder27** [comment]
If only Lua was under the GPL... They would have to be able to give us object files, for our linking convenience.

**Comment by polarathene** [comment]
So you're doing something quite similar to how CheatEngine manipulates an applications memory? Just you're doing it via things like IPHLPAPI? I'm no help here so I'll leave you guys to it, best of luck :)

If it helps any, when I was googling to see how others were doing lua hooks for games, I think someone mentioned taking advantage of the steam api that the game called externally. No idea if that's the STEAM:some_func() calls I see in lua, might help you pinpoint a lua function though? *shrugs*

**Comment by minermoder27** [comment]
I've just found https://github.com/gaffe23/linux-inject
We can't use this when running in a different process, but we should be able to run it in glibc, after overriding it with LD_PRELOAD, so we can run it in the context of PAYDAY2. After that, we could call PAYDAY2's Lua functions.

**Comment by SirWaddles** [comment]
It looks like that could work, but I don't know about having to get players to disable important memory protection features just to get the thing to work.

**Comment by minermoder27** [comment]
Solution: we don't.

On many Linux distributions, the kernel is configured by default to prevent any process from calling ptrace() on **another process that it did not create** (e.g. via fork()).

Therefore, run this from within glibc, from within the process itself. As a process can modify itself, we can leave this on.

**Comment by SirWaddles** [comment]
I'm kind of confused.

Isn't the point of that library so you can get inside another process space? If you're already there, why do you need it?

**Comment by minermoder27** [comment]
Actually, valid point. I think I was over-thinking it quite a lot :P

**Comment by SirWaddles** [comment]
I guess if you hook in via LD_PRELOAD, you don't even need to proxy anything. Since there's no point wrapping any of the shared functions (they're useless to us) the only problem left to solve is being in the space, and we're there anyway.

We know the addresses of every function, thanks to the symbol exports, so all that's left is we need to find a way to neatly start overriding, and calling functions based on those symbol addresses.

**Comment by minermoder27** [comment]
ptrace()

see man ptrace
or http://www.linuxjournal.com/article/6210 as an example of injecting code into a running program.
Pretty much exactly what we want.

**Comment by BangL** [comment]
i dont understand any word, but it looks kinda promising. if you need a tester, let me know^^

**Comment by minermoder27** [comment]
TL;DR of the conversation:
We've found a way (that should work) to get our code into PAYDAY2. I'm currently writing a test program, though a lot more work is probably needed to actually get things to the point where you can drag in a mod and have it work.

**Comment by goontest** [comment]
Found this
https://github.com/RomanHargrave/blt4l

**Comment by minermoder27** [comment]
Okay, found https://github.com/Zeex/subhook
Very simple way to override methods.

```
typedef void (*foo_func)(int x);
subhook_t foo_hook;

void my_foo(int x) {
  printf("foo(%d) called\n", x);

  /* Call foo() via trampoline. */
  ((foo_func)subhook_get_trampoline(foo_hook))(x);
}

int main() {
  /* Create a hook that will redirect all foo() calls to to my_foo(). */
  foo_hook = subhook_new((void *)foo, (void *)my_foo);

  /* Install it. */
  subhook_install(foo_hook);

  foo(123);

  /* Remove the hook and free memory when you're done. */
  subhook_remove(foo_hook);
  subhook_free(foo_hook);
}
```

**Comment by ZNixian** [comment]
@SirWaddles Got Lua injection working with the Lua command, so I can run arbitrary code within the context of the Lua command, and also presumably PAYDAY 2. As such, all that is left is to port the non-injection parts of BLT to GNU/Linux, which should be comparatively straightforwards.

**Comment by SirWaddles** [comment]
Yea, most of that is cross platform

**Comment by karolherbst** [comment]
I could also help to work on this, because I am a Linux dev and own a copy of payday2 :)

**Comment by SirWaddles** [comment]
Up to you. There's already that port that @goontest mentioned, and I was under the impression that @ZNixian was working on a PR? Not sure.

Almost seems like this one is out of my hands.

**Comment by karolherbst** [comment]
well I don't care who does it in the end, because I have usually enough to do with nouveau stuff. But I will test it in the end and most likely try to change stuff I don't like

**Comment by ZNixian** [comment]
What the problem is at the moment (not helped by me never having used x86 assembly before. If you do know assembly, then it would be amazing if you could help. The problem is, when I am trying to inject into PD2 from a shared library (NOT using PIC, by the way - it would break everything). However, the SubHook library I am using only uses a relative jmp command - so it ONLY works if PD2 is loaded within 2GB of the Lua functions we need.

If you could fork subhook, and make it at least kinda work with 64bit absolute jumps, that would make BLT for Linux possible.

**Comment by karolherbst** [comment]
You don't want to use subhook, because this opens a world of pain.

In the end you want to just intercept the right call to an external library, everything else you can't do reliable.

Now in fact you want to use maybe LD_PRELOAD and just overwrite the thing payday2 calls to initialize LUA, dlopen the real thing (can also be the currently running binary with dlopen(NULL)) and dlsym the real function.

This is the less painfull thing to overwrite external functions.

**Comment by ZNixian** [comment]
Unfortunately, a copy of Lua is statically linked to PD2. The reason why I was using subhook was I don't know assembly, and it looked like a easy way to do things. If there is a better way, by all means please give me an example I can use.

**Comment by Ozymandias117** [comment]
I finally found some time to try to help with the hooking situation.
I've made a temporary fork of subhook which pushes the address onto the stack and returns rather than uses a jump: https://github.com/Ozymandias117/subhook

See if this works any better for you. Note that I would advise against using the trampoline, and would instead uninstall the hook prior to calling the function and reinstalling it when you're done. This will be slightly less efficient, but should be safer due to the large memory space.

**Comment by LeonardKoenig** [comment]
Hi all, x-posting from here https://github.com/RomanHargrave/blt4l/issues/3#issuecomment-205339411 and here https://github.com/RomanHargrave/blt4l/issues/5:

I've made it to override "dsl::EventManager::update()", but sadly when calling the original function from within the hooked one it hangs while trying to `std::__1::recursive_mutex::lock()` -- so we have a multi threading problem. I've no idea about multi threading though so help would be appreciated if you have ideas on how to solved this.

Note: I use @Ozymandias117 subhook version since the other one completely segfaults and doesn't even hook for some reason and you need to un-comment the lines for Installing the subhook in `src/hook.cc` if you want to 'activate' the feature.

Just compile as described in the README and run:
```
export SteamAppId=218620
export LD_PRELOAD="/path/to/libblt_loader.so $LD_PRELOAD"
"/path/to/steamapps/common/PAYDAY 2/payday2_release"
```

Note that because of the way LD_PRELOAD works, no spaces are allowed in the hook-libs path.

For debugging with gdb you can automate this too (call with `gdb -x gdbinit.txt`)
```
set environment SteamAppId=218620
set environment LD_PRELOAD="/path/to/libblt_loader.so $LD_PRELOAD"
file "/path/to/steamapps/common/PAYDAY 2/payday2_release"
```

**Comment by RomanHargrave** [comment]
To follow up what Leonard said, and from what I have seen of BLT4Win (gah! WINAPI! kill it with fire), I don't know that we need to be _terribly_ caught up with multi-threading concerns as long as we aren't too intrusive WRT the stack, and because for each thread, there is a different LUA context, and we should logically always get the correct LUA context for a thread as a parameter.

Forgive me if I've skipped over or misstated something. I've just gotten off an eight hour train ride.

**Comment by RomanHargrave** [comment]
@JamesWilko could you perhaps tell me which function you _think_ `do_game_update` would be, symbolically? I'm having a little trouble figuring out what it is that you're calling, and I would like for optimal compatibility.

**Comment by SirWaddles** [comment]
I'm not sure really, but it seems to run every frame and also has a pointer to the lua state in its class owner and seems to take eight bytes of data.

Originally, we were using it to get the lua state, but we moved off of that later on. Now it's just being used to call lua stuff synchronously (to tell Payday that the contents of a HTTP request have been receieved, on its own thread)

You could probably just pause Payday in a debugger at any point and it'll be in the callstack somewhere. I'm guessing it'd be called 'Update' or 'Tick' or something like that.

**Comment by RomanHargrave** [comment]
Then it's probably Application::Update(). But that, nor do many other functions like it take a parameter, other than probably `this`, which I assume you are referring to.

**Comment by RomanHargrave** [comment]
OK. 

It's definitely `Application::update(this)`.

I was just stupid, and being a C programmer, didn't think about stupid parameter nonsense. I wish C++ still forced you to insert "this" in to the function parameters like the "good" old days.

**Comment by Ozymandias117** [comment]
I spent a little while last night trying to port this one over. It builds and hooks (mostly) the same functions...
The issues currently are:
Console was just removed for now (It didn't appear to be too important at the time)
lua_newstate is hooked instead of luaL_newstate. The latter did not appear to ever get called on Linux
Other than the console output, mods aren't loading, so there are obviously issues remaining.

I created a basic Makefile for now, and once we get it working, there will be effort involved in having both building - there were several places where msvc and gcc disagreed on "valid" code. You can see the current progress on this branch: https://github.com/Ozymandias117/Payday-2-BLT/tree/linux-work if you have any insights.

Build instructions from checkout directory:
git submodule update --init
mkdir ./build
make

It requires the -dev files for libcurl and openssl. (On Debian I've been building with: libcurl4-gnutls-dev, libssl-dev, and presumably build-essential covers anything else)

**Comment by karolherbst** [comment]
sounds good, I will give it a shot tomorrow. Maybe there can be also something be used like systemtap to see what kind of functions are called (for development)? I am sure you can just hook all lua* functions with it, but never used it myself.

**Comment by RomanHargrave** [comment]
@karolherbst We shouldn't need to hook more than two LUA functions, according to BLT4WIN (this project). Seeing as the LUA library is pretty platform-agnostic, I'm proceeding with the assumption that a lot of this code can just be ported.

**Comment by karolherbst** [comment]
@RomanHargrave yeah I know. I just meant this way a developer could just trace what PD2 is doing on the LUA side

**Comment by Ozymandias117** [comment]
@RomanHargrave Are we sure about that? Two of the functions that are hooked with Detours, lua_call and lua_close take the same arguments as the original functions. do_game_update takes some values... but I don't know how those were decided. The only one that seems different from what I expect is luaL_newstate, although I don't know why it's different.

According to the lua docs, it seems like luaL_newstate doesn't take any parameters? However in CREATE_CALLABLE_CLASS_SIGNATURE it's shown as taking (char, char, int), then in lua_newstate_new it's taking (void*, int, char, char, int)... That shouldn't even be a class with a this pointer? Both sides also show it returning an int when the documentation says it returns a lua_State*...
(For reference: http://www.lua.org/manual/5.1/manual.html#luaL_newstate)

@SirWaddles Do you happen to have any input as to why luaL_newstate is being hooked the way it is?

**Comment by SirWaddles** [comment]
Yea, PD2s lua code is not 1:1 with the lua library. They changed the luaL_newstate function so that they could add their own lua functions and stuff as it gets initalized. We discovered this when we were disassembling the binary.

**Comment by RomanHargrave** [comment]
@SirWaddles interesting.

I'll keep that in mind. I imagine that the changes carried to the linux release. WRT _newstate, I intend to hook `lua_newstate` since `luaL_newstate` calls it anyways.

**Comment by SirWaddles** [comment]
Small note, I'm not sure if it's a concern, but we tried that originally and it didn't work for us.

It was because PD2 defines its own pcall and dofile functions outside of lua_newstate, and then later removes them. We needed to grab the state *after* it did this black magic, because we needed to overwrite those functions in the lua code itself.

**Comment by Ozymandias117** [comment]
Interesting. I'll check again, but when I'd tried hooking luaL_newstate I never hit the hooked function, which was why I'd changed to lua_newstate in the first place. I'll go back and verify it was hooking at the right address, and that the memory at that address was modified.

**Comment by RomanHargrave** [comment]
@SirWaddles That's interesting, especially given what @Ozymandias117 has to say about luaL_newstate on linux...

**Comment by SirWaddles** [comment]
Thinking about it, that sounds right. I'd say the function we called luaL_newstate wasn't actually luaL_newstate in PD2, and was a function in Payday's lua wrapper somewhere with some code copied from luaL_newstate.

If there was some code that was hanging around, but never really got called, we probably wouldn't have paid much attention to it (i.e. the actual luaL_newstate)

**Comment by Ozymandias117** [comment]
That makes a lot more sense, and also explains why you're hooking it like it's a class rather than a pure c function.

I just verified again and saw that luaL_newstate was hooked properly, but didn't get called. I guess we need to figure out what the wrapped function is.

**Comment by Ozymandias117** [comment]
Here we go:
`(gdb) bt
#0  0x0000000000e0eed0 in lua_newstate ()
#1  0x0000000000dad948 in dsl::LuaInterface::newstate(bool, bool, dsl::LuaInterface::Allocation) ()
#2  0x0000000000daeafb in dsl::LuaInterpreter::LuaInterpreter(int, bool, bool, bool, dsl::LuaInterface::Allocation) ()
#3  0x0000000000bdd0a9 in Application::Application() ()
#4  0x0000000000e64e51 in app_run() ()
#5  0x0000000000e65e65 in app_entry_filtered() ()
#6  0x0000000000e64b74 in main ()
`

**Comment by SirWaddles** [comment]
`dsl::LuaInterface::newstate` sounds about right.

**Comment by Ozymandias117** [comment]
Although that still isn't really (char, char, int)?
I'm also still confused why your hooked function has an additional parameter that it's not passing on to the original function?

**Comment by SirWaddles** [comment]
Same again, all we had was the byte sizes. 1 byte we assumed was a character, 4 bytes an int etc.

The int edx param was as a result of using the __thiscall and __fastcall matched calling conventions.

**Comment by Ozymandias117** [comment]
Awesome, thanks for the information! I can load payday and get to the update manager, so we probably have working mods. I'll clean up and push to my branch, then start looking at the CConsole stuff.

**Comment by RomanHargrave** [comment]
@Ozymandias117 interesting work.

I'd like to look at merging it with my project, though, if that's not too big of a deal. We've already got the traction, plus, there are a few things that I'd like to clean up.

**Comment by Ozymandias117** [comment]
Yeah, I'm fine with that. Honestly, with as much as I changed here, we'll probably be able to get your hook working faster than I can figure out how to make both Windows and Linux build from the same tree...

**Comment by RomanHargrave** [comment]
@Ozymandias117 yeah, sounds like PDMods is going to need some work.

In the mean time, @SirWaddles @JamesWilko you may be interested in knowing that the demangled form of dsl::LuaInterface::newstate takes the params `(bool, bool, void*)`.

**Comment by Ozymandias117** [comment]
Also, @JamesWilko, will you have an issue with this change to the CREATE_CALLABLE_SIGNATURE's?

`name ## ptr name ## _orig = NULL; \`

When I was keeping the original name, I was unable to build with gcc.

Looking back through, as long as you're okay with that, it might not be that bad to get both building.

**Comment by SirWaddles** [comment]
Hey, as long as it still works on Windows, I don't really care. Not sure why gcc stops working though, seems weird.

**Comment by RomanHargrave** [comment]
@Ozymandias117 I'm getting some strange stuff coming down the line for dsl::LuaInterface::newstate 

Calling lua_gettop on the returned state structure is returning what I believe to be garbage data. Any thoughts?

Here are the parameter values for LuaInterface::newstate(..) as output by my debug code:

```
dsl_lua_newstate(this = 0x3de9810, 1, 0, 0x1)
stackSize = 663
dsl_lua_newstate(this = 0x3de98f0, 1, 1, 0)
stackSize = 1042594780
dsl_lua_newstate(this = 0xc0772c0, 1, 0, 0x1)
stackSize = -12637120
```

The `this` pointer appears relative consistent, and calling the original function with it works, the allocator portion is shaky, but appears to work too. What's getting me here are those bizarre

**Comment by Ozymandias117** [comment]
I just ran a Cook Off with my branch to test. Here's what I got:

Working:
Bodhi's War Repair
Fast.Net Standalone
Flashing Swan Song
HUDListblt
Lobby Player Info
Side Jobs in Lobby
SPL Sets

Didn't work:
Enhanced Hitmarkers
JackHUD
PocoHud

**Comment by RomanHargrave** [comment]
@Ozymandias117 I can't even get your port working. The hooks are working, but nothing else. I'm trying to confirm some of my stuff with it.

**Comment by SirWaddles** [comment]
What happened with PocoHUD? I know a lot of people use that one.

**Comment by RomanHargrave** [comment]
My guess is that those mods are using plaform-specific file paths, without looking at them.

**Comment by Ozymandias117** [comment]
@SirWaddles Here's the error output from Hitmarkers and Poco, although I doubt it's too relevant here specifically:

Enhanced Hitmarkers:
`[EH] Creating Enhanced Hitmarkers overrides folder...
CURL INITD
sh (27916)
sh: 1: Syntax error: Unterminated quoted string
assets/mod_overrides/Enhanced Hitmarkers/guis/textures/pd2/
In GetDirContents
mods/Enhanced Hitmarkers/hitmarkers/
In GetDirContents
loverkill.texture isDir 0
checkbox K.texture isDir 0
star.texture isDir 0
. isDir 1
pd2.texture isDir 0
almir.texture isDir 0
smiley.texture isDir 0
pirate H.texture isDir 0
aperture.texture isDir 0
nope.texture isDir 0
horny.texture isDir 0
pirate K.texture isDir 0
ovk.texture isDir 0
phatcross.texture isDir 0
tongue and lip.texture isDir 0
hotline miami.texture isDir 0
pooverkill.texture isDir 0
classic crit.texture isDir 0
Malo.texture isDir 0
checkbox H.texture isDir 0
classic hit.texture isDir 0
TdlQ.texture isDir 0
pattee.texture isDir 0
radioactive.texture isDir 0
`

PocoHud
`Err: Failed to load mods/PocoHud3/poco/../poco/3rdPartyLibrary.luacPoco:Init
Err: Failed to load mods/PocoHud3/poco/../poco/Hud3_class.luacErr: Failed to load mods/PocoHud3/poco/../poco/Hud3_Options.luacLocale EN loaded
Err: Failed to load mods/PocoHud3/poco/../poco/Hud3.luacIncoming HTTP Request/Request
http://api.paydaymods.com/updates/retrieve/?mod[0]=payday2blt&mod[1]=payday2bltdll&mod[2]=pocohud3&mod[3]=SDJBL&mod[4]=fastnetstand&mod[5]=SPLS&mod[6]=LPI
814
`

JackHUD
`mods/JackHUD/Lua/CrimeNetFiltersGui_ext.lua:7: attempt to call method 'save_persistent_settings' (a nil value)
`

@RomanHargrave Let me build again from fresh and make sure I didn't miss a file in the repo or anything. The two things I can think of to suggest are: 1. Did you copy the mods folder into the Payday 2 folder? 2. Currently I've only been able to get it working when running from the command line. If I add it to Steam's launch options, it didn't work.

**Comment by RomanHargrave** [comment]
@Ozymandias117 Got it working; not having issues with PocoHUD though.

**Comment by Ozymandias117** [comment]
Hmm. I'll try downloading a fresh copy of Poco.

**Comment by RomanHargrave** [comment]
@Ozymandias117 btw, got my newstate hook working, was not expecting the supposed 'this' pointer to be a lua_state** value

**Comment by SirWaddles** [comment]
Yea, the lua_state is just the first member of the class. So `this` points to the class, derefence it to get the first member, you've got a pointer to the lua state. It's technically not a lua_state**, but it's close enough.

**Comment by Ozymandias117** [comment]
Even with a fresh Poco, I see it under Options -> Mod Options, but backspace doesn't work to open up the settings menu, and I'm getting those errors.

**Comment by RomanHargrave** [comment]
@SirWaddles I figured as much.

**Comment by SirWaddles** [comment]
I didn't realise Poco was luac files? Could that be the problem?

**Comment by Ozymandias117** [comment]
GoonMod required me to rename req/options.lua to req/boptions.lua... it was loading updates.lua prior to options.lua for some reason... I'm not sure how that's working on Windows...?

Looking into JackHUD next

**Comment by Ozymandias117** [comment]
@SirWaddles It looks like they search for the correct name, then if it doesn't exist, or dofile fails, they try with c appended. There's no output in the case that it finds the original file, but dofile fails.

**Comment by LeonardKoenig** [comment]
@SirWaddles Poco is distributed as binary luac buy lua code is available on GitHub too, so one could at least debug it.

**Comment by BangL** [comment]
@Ozymandias117 
"It turns out it was named NetworkMatchmakingSTEAM_ext.lua in the folder."

..i've already changed that locally, but git on windows doesnt care.
thanks for pointing it out though. i will take care of it.

**Comment by BangL** [comment]
btw.. this is awesome news. cant wait to get home. love you guys.

**Comment by karolherbst** [comment]
@Ozymandias117 awesome, your branch works for me (after I also copied the mods folder). But for me it wants to update the blt DLL, just curious if you also got the message.

**Comment by Ozymandias117** [comment]
@karolherbst Yeah, that still needs to be fixed.

**Comment by SirWaddles** [comment]
Yea, if it can ask for a different mod ident instead of `payday2bltdll` like `payday2bltdll_linux` or something, I can add it to the website.

**Comment by BangL** [comment]
about pocohud:

the errors about luac files can be understood as warnings.
as you can see in common.lua, line 18,  pocohud tries to run luac files IF lua versions are not existant:
`return __req(name) or __req(name..'c')`

the thing with the backspace key is a different story... for me its not working as well, but if i tap my X key to times, the options are coming up. (which is my binding for the detail view, i think it is the tilde-key by default)

**Comment by karolherbst** [comment]
well I tried pocohud and it seems work just fine, except I couldn't configure it.

**Comment by BangL** [comment]
> well I tried pocohud and it seems working just fine, except I couldn't configure it.

thats what i tried to explain in my post before.
the errors arent errors.
and the configuring isnt possible because of messed up keybindings. - backspace isnt working as intended for some reason.
but if you tap the "detail view" key two times, you can access the settings, which is the tilde-key by default.

**Comment by Ozymandias117** [comment]
Yeah, I'd posted earlier about the or in common.lua. My concern is why are we getting into that or case? It implies to me a bug in luaF_dofile. My guess is that it shouldn't be returning 0?
On Apr 7, 2016 10:29 AM, Henno Rickowski <notifications@github.com> wrote:
well I tried pocohud and it seems working just fine, except I couldn't configure it.


thats what i tried to explain in my post before.
the errors arent errors.
and the configuring isnt possible because of messed up keybindings. - backspace isnt working as intended for some reason.
but if you tap the "detail view" key two times, you can access the settings, which is the tilde-key by default.

**Comment by BangL** [comment]
> Yeah, I'd posted earlier about the or in common.lua. My concern is why are we getting into that or case?
indeed. you're absolutely right. i didn't even think about that so far.
> It implies to me a bug in luaF_dofile. My guess is that it shouldn't be returning 0?
makes perfect sense, yea.   the "or" case will only happen when the statement in front of it is nil, false or 0.
so luaF_dofile should return 1 if it ran properly, right.

**Comment by BangL** [comment]
@Ozymandias117 
just debugged it.. actually dofile() isnt even returning 0, it should, but it is returning nil yet. - thats the point.

**Comment by RomanHargrave** [comment]
I just got my fresh implementation of BLT working. No issues so far, other than the current pocohud issue (and multiple poco components like damage and buff indicators are fine over here).

**Comment by Ozymandias117** [comment]
Nice! I'm out of town until Monday - at that point I can try to finish merging this one back and help with any other bugs that have been found.

**Comment by SirWaddles** [comment]
This is largely my fault, but that might be an issue as the release branch at the moment is actually the crimefest branch.

**Comment by RomanHargrave** [comment]
@SirWaddles what exactly is different in the crimefest branch?

**Comment by SteamD** [comment]
awesome work folks. made a few notes while watching the thread, may or may not be useful at some point.

on ubuntu 14.04 you'll get an issue using gnutls, instead openssl is working fine.
`undefined symbol: CRYPTO_set_locking_call`
`aptitude remove libcurl4-gnutls-dev; aptitude install libcurl4-openssl-dev`

pocohud seems to be accessable via `k` too.
modified L18 in common.lua regarding the loading of compiled scripts (luac)
`--      return __req(name) or __req(name..'c')`
`        return __req(name)`

goonmod gets stuck on buying new inventory, modified mod.txt removed
`                { "hook_id" : "lib/managers/menu/blackmarketgui", "script_path" : "goonbase.lua" },`
`                { "hook_id" : "lib/managers/blackmarketmanager", "script_path" : "goonbase.lua" },`

**Comment by Ozymandias117** [comment]
@SteamD The PocoHud change is unnecessary, and may break PocoHud for some current users based off of @LeonardKoenig's statement that it's distributed as luac somewhere. The error output is erroneous. I don't believe anyone has solved the backspace keybinding issue yet.

**Comment by BangL** [comment]
> I don't believe anyone has solved the backspace keybinding issue yet.

pocohud is using `Idstring('backspace')` to handle that binding
so i bet it's a linux-related problem on overkills side (just a very wild guess)

**Comment by drewofdoom** [comment]
@Ozymandias117 @SteamD @LeonardKoenig The only reference to "luac" at all in the actual code is in a comment in Hud3.lua. Just went through and grepped all of the lua files for it.

The original, non-BLT version is compiled as luac, so that may be causing the confusion.

**Comment by BangL** [comment]
@drewofdoom 
it's this:  `__req(name..'c')`

**Comment by SirWaddles** [comment]
@RomanHargrave Overkill changed the way their lua implemenation works to use __fastcalls and stuff. I don't think it'll affect the linux one, but the difference is we have to do some black magic assembly to restore the stack on each call.

**Comment by RomanHargrave** [comment]
@SirWaddles yeah, Linux is having no issue using a sensible calling convention.

**Comment by SirWaddles** [comment]
I just meant it as a note if anyone wants to PR into upstream, the weirdness calling stuff needs to stay for the windows build, but at the moment it's only on the crimefest branch. @saltisgood recently fixed it up so it was tidied into the signature.h macros, so it should be simple to just replace those in an ifdef.

**Comment by RomanHargrave** [comment]
@SirWaddles ah, I see.

Yeah, there aren't any Linux compilers that support that convention, and as far as I know it's more just a leftover MSVC feature from the days of old. No idea why OVK is using it.

**Comment by SirWaddles** [comment]
I'm not completely certain that they are, the changes we noticed that there were a lot of inlined lua functions now, and a lot of stuff being passed through registers. We added __fastcall to the signature.h macro, did the black magic assembly and it seemed to fix our post-crimefest release.

We got it working before the HoxHUD team, but I am curious how they went about the problem.

**Comment by LeonardKoenig** [comment]
Well, `__fastcall` tries to pass arguments in registers but is ignored (by VC) when compiling for amd64. So I suppose we are lucky here since the Linux build is 64bit and I suppose that GCC does the same on Linux.  
Also I could imagine that `__fastcall` is done automatically as an optimization, but cannot be sure of course.

**Comment by Ozymandias117** [comment]
Sorry, I've been out of town.

With the change to move the mods into a separate repo, and from what I read on @RomanHargrave's issue tracker, is it still worth trying to get both building here, or should we just keep the dll/so hooking code separate?

**Comment by SirWaddles** [comment]
There are more utility functions I'd like to add to the build, things that can't necessarily be done in lua.

If you're happy to maintain those functions in the linux version, that's fine. I don't mind writing cross-platform code, I'm just not gonna write two versions of the same thing.

**Comment by RomanHargrave** [comment]
@Ozymandias117 I think that it would be more worthwhile if anyone that is willing to put work in to a modding system for Linux concentrated on one project, rather than two. You should already have organization-level access to blt4linux, so you are welcome to come over and help us out. 

@SirWaddles what functionality were you thinking of implementing?

**Comment by SirWaddles** [comment]
We've had a few ideas in the past, for example I wanted to build a means for different cliens to send messages back and forth without necessarily being in the same lobby. This was largely for the trading in GoonMod to work outside of a lobby.

Other things have been different kinds of audio playback and generation, I also briefly looked at an in-game mod manager (i.e. one that you can actually download new mods while in-game) that would've required some custom rendering.

Overall I'm not sure, we don't have a formal milestone plan or anything, but I'd like to keep the option open for new features.

**Comment by RomanHargrave** [comment]
@SirWaddles perhaps since we have two mods using the same LUA base, and hence exposing the same API, we should put some sort of spec document together.

**Comment by SirWaddles** [comment]
It's basically just [the global functions in the docs](https://payday-2-blt-docs.readthedocs.org/en/latest/globals/functions/) really, as well as the 'file' library.

**Comment by karasuhebi** [comment]
So where's this at nowadays? Any hope?

**Comment by polarathene** [comment]
@karasuhebi Yes you can find it at: https://github.com/blt4linux/blt4l

**Comment by RomanHargrave** [comment]
@karasuhebi it works 100%. As long as a BLT mod doesn't make any assumptions about the file system (like using \ instead of / or using drive letters), it will work on Linux.

**Comment by polarathene** [comment]
@JamesWilko There is no mention of the linux build/repo on the README of this repo or on the official [BLT site download page](http://paydaymods.com/download/). Closing the issue decreases that visibility more(although makes sense). Could you at least link/mention the linux repo in the README?
